numpy/lib/recfunctions.py

```diff
@@ -972,6 +1018,30 @@ def structured_to_unstructured(arr, dtype=None, copy=False, casting='unsafe'):
                                  'itemsize': arr.dtype.itemsize})
     arr = arr.view(flattened_fields)
 
+    if (not copy) and all(dt.base == out_dtype for dt in dts):
+        # all elements have the right dtype already; if they have a common
+        # stride, we can just return a view
+        common_stride = _common_stride(offsets, counts, out_dtype.itemsize)
+        if common_stride is not None:
+            # ensure that we have a real ndarray; other types (e.g. matrix)
+            # have strange slicing behavior
+            arr = arr.view(type=np.ndarray)
+            new_shape = arr.shape + (sum(counts), out_dtype.itemsize)
+            new_strides = arr.strides + (abs(common_stride), 1)
```

Comment:
Why not use a negative `common_stride`? It does mean you have to be careful with the offset below, but there is nothing wrong with a negative stride (right now, your `[::-1]` at the end does the same thing).

Reply:
Can you give a suggestion of how you would do that cleanly?

I can of course do something like this:
```python
# first_field_offset is the first byte of the first field
ar1 = arr[..., first_field_offset:]
# new_strides may be negative here
arr2 = np.lib.stride_tricks.as_strided(arr, new_shape, new_strides)
```

If the stride is negative, all but the last element will be cut off by the slicing operation.
The `as_strided()` call then alters the strides, so that all elements can be accessed again.

I don't like that, because `arr2` is derived from `arr1`, but it is accessing memory that is out-of-bounds for `arr1`.

Reply:
Looking at it again, I think things should just work with a negative stride, as long as you make sure that the offset is guaranteed to be for the first item in the list - for a negative stride, that will be largest offset, so the slice below will not go out of memory.

It may well mean that you have to adjust `_common_stide` - but I would not be surprised if it actually simplified it!

Reply:
> Looking at it again, I think things should just work with a negative stride, as long as you make sure that the offset is guaranteed to be for the first item in the list - for a negative stride, that will be largest offset, so the slice below will not go out of memory.

You mean the code that I showed in my previous comment, but with `first_field_offset == offsets[0]`? That is what I meant, and I do not like that approach (see below).

> It may well mean that you have to adjust _common_stide - but I would not be surprised if it actually simplified it!

I don't see how that simplifies that function. It would stay identical as far as I can see.
Can you point me to what you would simplify or give an example?


<hr>

Let me illustrate why I don't like using a negative stride here:

Let's say we have a structured array with two fields 'a' and 'b'. The first field is 'a', but 'b' is the first field in memory.
This would look like this: (b0 is the first byte of b, b1 the second, etc. The array indices are written on top. They are equal to byte offsets at this point, because we are viewing the array as uint8.)

```
 0  1  2  3  4  5  6  7
[b0 b1 b2 b3 a0 a1 a2 a3]
```

Now we perform the slicing operation to set the correct offset. `first_field_offset` is 4 in this case.

```python
> arr = arr[..., first_field_offset:]
```

Now the memory looks like this:
All bytes of the field b are out of bounds.

```
 -  -  -  -  0  1  2  3  
[b0 b1 b2 b3 a0 a1 a2 a3]
```

Now we set the correct offset.
`new_shape` would be something like (..., 2, 4).
`new_strides` would be something like (..., -4, 1).

_This is the part I do not like_, because we are effectively accessing memory, that was (temporarily) out-of-bounds.

```python
> arr = np.lib.stride_tricks.as_strided(arr,
                                        new_shape,
                                        new_strides)
```

After that, it looks like this:

```
 (1, 0) (1, 1) (1, 2) (1, 3) (0, 0) (0, 1) (0, 2) (0, 3)
[b0     b1     b2     b3     a0     a1     a2     a3    ]
```